File: dtc-flink/src/main/scala/com/dtc/analytic/java/demo/KafkaMessageStreaming.java
//package com.dtc.analytic.java.demo;
//
//import org.apache.flink.api.java.tuple.Tuple;
//import org.apache.flink.api.java.tuple.Tuple2;
//import org.apache.flink.streaming.api.TimeCharacteristic;
//import org.apache.flink.streaming.api.datastream.DataStream;
//import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
//import org.apache.flink.streaming.api.functions.windowing.WindowFunction;
//import org.apache.flink.streaming.api.windowing.time.Time;
//import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
//import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer09;
//import org.apache.flink.streaming.util.serialization.SimpleStringSchema;
//import org.apache.flink.util.Collector;
//
//import java.util.Properties;
//
///**
// * Created on 2019-06-06
// *
// * @author :hao.li
// */
//public class KafkaMessageStreaming {
//
//    public static void main(String[] args) throws Exception {
//        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
//        env.enableCheckpointing(5000); // 非常关键，一定要设置启动检查点！！
//        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
//
//        Properties props = new Properties();
//        props.setProperty("bootstrap.servers", "10.3.6.7:9092,10.3.6.12:9092,10.3.6.16:9092");
//        props.setProperty("group.id", "bigdata_test");
//
//        FlinkKafkaConsumer09<String> consumer =
//                new FlinkKafkaConsumer09<>("connect-test", new SimpleStringSchema(), props);
//
//        consumer.assignTimestampsAndWatermarks(new MessageWaterEmitter());
//
//        DataStream<Tuple2<String, Long>> keyedStream = env
//                .addSource(consumer)
//                .flatMap(new MessageSplitter())
//                .keyBy(0)
//                .timeWindow(Time.seconds(10))
//                .apply(new WindowFunction<Tuple2<String, Long>, Tuple2<String, Long>, Tuple, TimeWindow>() {
//                    @Override
//                    public void apply(Tuple tuple, TimeWindow window, Iterable<Tuple2<String, Long>> input, Collector<Tuple2<String, Long>> out) throws Exception {
//                        long sum = 0L;
//                        int count = 0;
//                        for (Tuple2<String, Long> record: input) {
//                            sum += record.f1;
//                            count++;
//                        }
//                        Tuple2<String, Long> result = input.iterator().next();
//                        result.f1 = sum / count;
//                        out.collect(result);
//                    }
//                });
//
//        keyedStream.writeAsText("conf/test.txt");
//        env.execute("Flink-Kafka demo");
//    }
//}

File: dtc-flink/src/main/scala/com/dtc/analytic/scala/works/KafkaMessageStreaming.scala
package com.dtc.analytic.scala.works


import com.dtc.analytic.scala.common.DtcConf
import net.minidev.json.JSONObject
import net.minidev.json.parser.JSONParser
import org.apache.flink.api.common.functions.{FlatMapFunction, MapFunction, ReduceFunction}
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.util.Collector

/**
  * Created on 2019-06-06
  *
  * @author :hao.li
  */
object KafkaMessageStreaming {
  def main(args: Array[String]): Unit = {
    DtcConf.setup()
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    //    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
    //    env.setParallelism(1)
    //    val conf = DtcConf.getConf()
    //    val level = conf.get("log.level")
    //    val lev: Int = LevelEnum.getIndex(level)
    //
    //    val brokerList = conf.get("flink.kafka.broker.list")
    //    val topic = conf.get("flink.kafka.topic")
    //    val groupId = conf.get("flink.kafka.groupid")
    //    val prop = new Properties()
    //    prop.setProperty("bootstrap.servers", brokerList)
    //    prop.setProperty("group.id", groupId)
    //    prop.setProperty("topic", topic)
    val flieStream = env.readTextFile("/Users/lixuewei/workspace/DTC/dtc-analytics_0.3/dtc-flink/src/main/resources/long_text_2019-07-22-11-57-56.txt")

    //    val myConsumer = new FlinkKafkaConsumer09[String](topic, new SimpleStringSchema(), prop)
    //    val waterMarkStream = myConsumer.assignTimestampsAndWatermarks(new AssignerWithPeriodicWatermarks[(String)] {
    //      var currentMaxTimestamp = 0L
    //      var maxOutOfOrderness = 10000L // 最大允许的乱序时间是10s
    //      val sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    //
    //      override def extractTimestamp(element: (String), previousElementTimestamp: Long) = {
    //        val event = element.split("\\$\\$")
    //        var time1 = event(0).replace("T", " ").split("\\+")(0).replace("-", "/")
    //        var time2 = new Date(time1).getTime
    //        currentMaxTimestamp = Math.max(time2, currentMaxTimestamp)
    //        time2
    //      }
    //
    //      override def getCurrentWatermark = new Watermark(currentMaxTimestamp - maxOutOfOrderness)
    //
    //    })
    //    val text = env.addSource(flieStream)
    //    var inputMap = flieStream.flatMap(new myFlatMapFunction).keyBy("System_name").keyBy("Host_ip").sum("value")
    var inputMap = flieStream.flatMap(new myFlatMapFunction).keyBy("System_name").keyBy("Host_ip").reduce(new ReduceFunction[DataStruct] {
      override def reduce(t: DataStruct, t1: DataStruct): DataStruct = {
        var test = t.value + t1.value
        DataStruct(t.System_name, t.Host_ip, "cup", t.time, test)
      }
    })
    inputMap.print()
    //    val result = inputMap.filter(b => b._1.equals("linux")).filter(a => a._3.equals("a") || a._3.equals("b")).keyBy(1).sum(4)
    //    var result = inputMap.keyBy(2).sum(4)
    //    result.print()

    //    var window = inputMap.keyBy(0)
    //      .timeWindow(Time.seconds(10), Time.seconds(5))
    env.execute("StreamingWindowWatermarkScala")
  }

  class myFlatMapFunction extends FlatMapFunction[String, DataStruct] {
    override def flatMap(t: String, collector: Collector[DataStruct]): Unit = {
      var message: DataStruct = null
      val jsonParser = new JSONParser()
      val json = jsonParser.parse(t).asInstanceOf[JSONObject]
      val name = json.get("name").toString.split("_", 2)
      val system_name = name(0).trim
      val zhibiao_name = name(1).trim
      if (system_name.contains("node")) {
        val time = json.get("timestamp").toString
        val value = json.get("value").toString.toDouble
        val lable = jsonParser.parse(json.get("labels").toString).asInstanceOf[JSONObject]
        val lable_ip = lable.get("instance").toString.trim
        if ("a".equals(zhibiao_name)) {
          message = DataStruct(system_name, lable_ip, zhibiao_name, time, value)
          collector.collect(message)
        } else if ("b".equals(zhibiao_name)) {
          message = DataStruct(system_name, lable_ip, zhibiao_name, time, value)
          collector.collect(message)
        } else if ("c".equals(zhibiao_name)) {
          message = DataStruct(system_name, lable_ip, zhibiao_name, time, value)
          collector.collect(message)
        } else if ("d".equals(zhibiao_name)) {
          message = DataStruct(system_name, lable_ip, zhibiao_name, time, value)
          collector.collect(message)
        } else {
          message = DataStruct(system_name, lable_ip, zhibiao_name, time, value * 0)
          collector.collect(message)
        }
      }

    }
  }


}

class MyMapFunction extends MapFunction[String, DataStruct] {
  override def map(line: String): DataStruct = {
    var message: DataStruct = null
    val jsonParser = new JSONParser()
    val json = jsonParser.parse(line).asInstanceOf[JSONObject]
    val name = json.get("name").toString.split("_", 2)
    val system_name = name(0).trim
    val zhibiao_name = name(1).trim
    if (system_name.contains("linux")) {
      val time = json.get("timestamp").toString
      val value = json.get("value").toString.toDouble
      val lable = jsonParser.parse(json.get("labels").toString).asInstanceOf[JSONObject]
      val lable_ip = lable.get("instance").toString.trim
      if ("a".equals(zhibiao_name)) {
        DataStruct(system_name, lable_ip, zhibiao_name, time, value)
        //        message.system_name=system_name
        //        message.lable_ip=lable_ip
        //        message.zhibiao_name=zhibiao_name
        //        message.time=time
        //        message.value=value
      }
    }
    //    val time = json.get("timestamp").toString
    //    val value = json.get("value").toString.toDouble
    //    val lable = jsonParser.parse(json.get("labels").toString).asInstanceOf[JSONObject]
    //    val lable_ip = lable.get("instance").toString
    //    val result = (name, lable_ip, time, value)
    return message
  }
}


//class OneMapFunction extends MapFunction[String,JSONObject]{
//  override def map(t: String): JSONObject = {
//    val jsonParser = new JSONParser()
//    val json = jsonParser.parse(t).asInstanceOf[JSONObject]
//    return json
//  }
//}
//
//case class DataStruct() {
//  var system_name: String = null
//  var lable_ip:String=null
//  var zhibiao_name:String= null
//  var time:String=null
//  var value = 0.0
//
////  def this(system_name: String,lable_ip:String,zhibiao_name:String,time:String, value: Double) {
////    this()
////    this.system_name = system_name
////    this.lable_ip = lable_ip
////    this.zhibiao_name=zhibiao_name
////    this.time = time
////    this.value=value
////  }
////
////  override def toString: String = "DataStruct{" + "system_name='" + system_name + '\'' + ", lable_ip=" + lable_ip + '}'
//}
